src/Stack/Script.hs

```diff
@@ -152,31 +158,110 @@ scriptCmd opts go' = do
+                    , moduleNameString mn
+                    , " appears in multiple packages: "
+                    , unwords $ map packageNameString pns'
+                    ]
+        return $ Set.unions pns `Set.difference` blacklist
+
+hashSnapshot :: RIO EnvConfig SnapshotCacheHash
+hashSnapshot = do
+    sourceMap <- view $ envConfigL . to envConfigSourceMap
+    let wc = whichCompiler $ smCompiler sourceMap
+    compilerInfo <- getCompilerInfo wc
+    let maybePliHash dep | PLImmutable pli <- dpLocation dep =
+                             Just $ immutableLocShaBs pli
+                         | otherwise = Nothing
+        locHashes = mapMaybe maybePliHash $ Map.elems (smDeps sourceMap)
+        hashedContent = compilerInfo:locHashes
+    return $ SnapshotCacheHash (SHA256.hashLazyBytes $ BL.fromChunks hashedContent)
+
+mapSnapshotPackageModules :: RIO EnvConfig (Map PackageName (Set ModuleName))
+mapSnapshotPackageModules = do
+    sourceMap <- view $ envConfigL . to envConfigSourceMap
+    installMap <- toInstallMap sourceMap
+    (_installedMap, globalDumpPkgs, snapshotDumpPkgs, _localDumpPkgs) <-
+        getInstalled installMap
+    let globals = dumpedPackageModules (smGlobal sourceMap) globalDumpPkgs
+        notHidden = Map.filter (not . dpHidden)
+        notHiddenDeps = notHidden $ smDeps sourceMap
+        installedDeps = dumpedPackageModules notHiddenDeps snapshotDumpPkgs
+        dumpPkgs = Set.fromList $ map (pkgName . dpPackageIdent) snapshotDumpPkgs
+        notInstalledDeps = Map.withoutKeys notHiddenDeps dumpPkgs
+    otherDeps <- for notInstalledDeps $ \dep -> do
+        gpd <- liftIO $ cpGPD (dpCommon dep)
+        Set.fromList <$> allExposedModules gpd
+    -- source map construction process should guarantee unique package names
+    -- in these maps
+    return $ globals <> installedDeps <> otherDeps
+
+dumpedPackageModules :: Map PackageName a
+                     -> [DumpPackage]
+                     -> Map PackageName (Set ModuleName)
+dumpedPackageModules pkgs dumpPkgs =
+    let pnames = Map.keysSet pkgs `Set.difference` blacklist
+    in Map.fromList
+           [ (pn, dpExposedModules)
+           | DumpPackage {..} <- dumpPkgs
+           , let PackageIdentifier pn _ = dpPackageIdent
+           , pn `Set.member` pnames
+           ]
+
+mapMutablePackageModules :: RIO EnvConfig (Map ModuleName (Set PackageName))
+mapMutablePackageModules = do
+  deps <- view $ envConfigL . to envConfigSourceMap . to smDeps
+  fmap (Map.fromListWith mappend . concat) $ for (Map.toList deps) $ \(pname, dep) -> do
+      case dpLocation dep of
+          PLImmutable _ -> pure []
+          PLMutable _ -> do
+              gpd <- liftIO $ cpGPD (dpCommon dep)
+              modules <- allExposedModules gpd
+              return [ (m, Set.singleton pname) | m <- modules ]
+
+allExposedModules :: PD.GenericPackageDescription -> RIO EnvConfig [ModuleName]
+allExposedModules gpd = do
+  -- FIXME add os, arch conditionals
```

Comment:
Can you clarify this `FIXME`? It looks like you've already implemented these conditionals correctly.

Reply:
oops